Add a CMS Starter example

triozer opened the pull request:
### Description
This pull request introduces a new CMS starter example, including updates to various configuration files, data sources, and UI components. The changes aim to set up a basic CMS plugin for Framer, providing a template for further development.

### QA
- [ ] Clone the repository.
- [ ] Navigate to the `examples/cms` directory.
- [ ] Install dependencies and start the plugin locally.
- [ ] Adding a new Collection
  - [ ] Create a new collection.
  - [ ] Open the CMS plugin.
    - [ ] Confirm a prompt appears asking to select a collection to sync.
    - [ ] Verify you can select from the available collections.
- [ ] Managing an existing Collection
  - [ ] Open an existing collection that has been set up previously.
  - [ ] Click the "Manage" button.
    - [ ] Confirm the "Field Mapping UI" is displayed.
    - [ ] Verify you can:
      - [ ] Change the slug field to a valid string field.
      - [ ] Rename fields.
      - [ ] Toggle fields on/off by clicking them.
- [ ] Syncing an existing Collection
  - [ ] Open an existing, pre-configured collection.
  - [ ] Click the "Sync" button.
    - [ ] Confirm the sync process starts immediately.
    - [ ] Verify the collection syncs with the configured data source without further setup.

### Additional Checks
- [ ] Ensure all UI elements are properly styled and responsive.
- [ ] Confirm error states are handled gracefully.
- [ ] Verify field changes persist after saving.
- [ ] Ensure sync operations display appropriate loading states.

alecmev commented on examples/cms-starter/.gitignore:
Most (all) of this should be covered by the root `.gitignore`, no?

triozer replied:
Yes!

alecmev commented on examples/cms-starter/README.md:
````diff
@@ -0,0 +1,19 @@
+# Framer Plugin Template
+
+This is a template for using the Framer Plugin API in a TypeScript project.
+
+## Quickstart
+
+First, run the development server:
+
+```bash
+npm run dev
+# or
+yarn dev
+# or
+pnpm dev
+# or
+bun dev
+```
+
+Learn more: https://www.framer.com/developers/plugins/introduction
````
Nit: This is @huntercaron's wheelhouse, but, just in case, I'd maybe simplify or even eliminate the first paragraph, as we probably won't be maintaining a non-TypeScript version of this, and the rest is mostly a rehash of the title. And add a direct link to how to open the plugin, to reduce friction just a tiny bit.

````suggestion
# Framer Plugin Template

Run the development server:

```bash
npm run dev
# or
yarn dev
# or
pnpm dev
# or
bun dev
```

And [open in Framer](https://www.framer.com/developers/plugins/quick-start#opening-in-framer).

Learn more: https://www.framer.com/developers/plugins/introduction
````

tom-james-watson commented on examples/cms-starter/datasources/categories.json:
```diff
@@ -0,0 +1,429 @@
+            "Color": "green"
+        },
+        {
+            "Title": "Category 1",
```
Is there any need to have these numbered categories? Are the above ones not sufficient?

triozer replied:
No! Let's clean this up 👀

tom-james-watson commented on examples/cms-starter/src/data.ts:
```diff
@@ -0,0 +1,293 @@
+export const PLUGIN_COLLECTION_SYNC_REFERENCE_KEY = `collectionSyncReference`
+export const PLUGIN_COLLECTION_SYNC_SLUG_KEY = `collectionSyncSlug`
+
+export type DataSourceFieldType =
+    | "string"
+    | "date"
+    | "image"
+    | "reference"
+    | "richText"
+    | "number"
+    | "boolean"
+    | "enum"
+    | "color"
```
Can we not get this from the types exported on framer-plugin? Don't want to need to maintain this, really.

tom-james-watson replied:
We have `CollectionField["type"]`

tom-james-watson replied:
And `ManagedCollectionField`

tom-james-watson commented on examples/cms-starter/src/App.tsx:
```diff
@@ -0,0 +1,160 @@
+            .filter(field => field.field && !field.source.ignored)
+            .filter(field => !field.reference || field.reference.destination !== null),
+        syncSlugFieldId
+    )
```
Should this have a return here?

triozer replied:
This is not a function so we can't really return, but `syncCollection` will call `framer.closePlugin` on success

alecmev commented on examples/cms-starter/src/utils.ts:
```diff
@@ -0,0 +1,49 @@
+export function assert(condition: unknown, ...msg: unknown[]): asserts condition {
+    if (condition) return
+
+    const e = Error("Assertion Error" + (msg.length > 0 ? ": " + msg.join(" ") : ""))
+    // Hack the stack so the assert call itself disappears. Works in jest and in chrome.
+    if (e.stack) {
+        try {
+            const lines = e.stack.split("\n")
+            if (lines[1]?.includes("assert")) {
+                lines.splice(1, 1)
+                e.stack = lines.join("\n")
+            } else if (lines[0]?.includes("assert")) {
+                lines.splice(0, 1)
+                e.stack = lines.join("\n")
+            }
+        } catch {
+            // nothing
+        }
+    }
+    throw e
```
Nit: Is this worth it? I know it's used in other plugins, but this is going to be a boilerplate. If it's worth it - use [`assert`](https://github.com/browserify/commonjs-assert) or [`is`](https://github.com/sindresorhus/is) maybe?

triozer replied:
Yeah, let's remove this completely, we don't really need it. And I think a better flow here is to prompt to the user using `framer.notify`

alecmev commented on examples/cms-starter/src/data.ts:
```diff
@@ -0,0 +1,293 @@
+            }
+
+            assert(
+                true,
```
Hm?

triozer replied:
🙈

alecmev commented on examples/cms-starter/index.html:
```diff
@@ -0,0 +1,14 @@
+      <meta charset="UTF-8" />
+      <link rel="icon" type="image/svg+xml" href="/icon.svg" />
+      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+      <title>Cms Starter</title>
```
Nit:

```suggestion
      <title>CMS Starter</title>
```

tom-james-watson commented on examples/cms-starter/src/data.ts:
```diff
@@ -0,0 +1,293 @@
+    return result
+}
+
+const CELL_BOOLEAN_VALUES = ["Y", "yes", "true", "TRUE", "Yes", 1, true]
+
+export function getFieldValue(field: FieldConfig, value: unknown) {
```
I think we don't need to demonstrate this kind of parsing logic, because that is just specific to the plugin. We should just read the values directly and have them be valid in the source data. Can add a comment to say that in a real world you'll want to have more complex logic.

tom-james-watson commented on examples/cms-starter/src/data.ts:
```diff
@@ -0,0 +1,293 @@
+        const itemId = generateHash(slug)
+        unsyncedItems.delete(itemId)
+
+        const fieldData: Record<string, unknown> = {}
```
```suggestion
        const fieldData: CollectionItemData["fieldData"] = {}
```

tom-james-watson commented on examples/cms-starter/src/data.ts:
```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+export function computeFieldConfig(existingFields: ManagedCollectionField[], dataSource: DataSource) {
```
Can we avoid the need for the complexity of this function? Can we just map the type of the source data more directly to our internal types? This is quite confusing for an example given all the custom types it requires.

huntercaron commented:
Lets call the actual folder "cms" instead of "cms-starter" since its already in a folder called "examples"

tom-james-watson commented on examples/cms-starter/src/App.tsx:
```diff
@@ -0,0 +1,137 @@
+    const [selectedDataSourceId, setSelectedDataSourceId] = useState<string | null>(
+        syncDataSourceId || allDataSources[0] || null
+    )
+    const [selectDataSource, setSelectDataSource] = useState<DataSource | null>(syncDataSource)
```
```suggestion
    const [selectedDataSource, setSelectedDataSource] = useState<DataSource | null>(syncDataSource)
```

tom-james-watson commented on examples/cms-starter/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+                        onChange={e => setSlugFieldId(e.target.value)}
+                        required
+                    >
+                        {possibleSlugFields.map(field => {
+                            if (!field) return null
+                            return (
+                                <option key={field.id} value={field.id}>
+                                    {field.name}
+                                </option>
+                            )
+                        })}
```
I can't select any fields for either of the two collections

triozer replied:
Fixed!

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,126 @@
+
+// Plugin keys
+export const PLUGIN_PREFIX = "cms_starter"
+export const LOCAL_STORAGE_LAST_LAUNCH_KEY = `${PLUGIN_PREFIX}.lastLaunched`
```
Is this used anywhere?

triozer replied:
Not anymore! Nice catch

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,142 @@
+} from "./data"
+import { FieldMapping } from "./FieldMapping"
+
+const activeCollection = await framer.getManagedCollection()
+const existingFields = activeCollection ? await activeCollection.getFields() : []
+
+const syncDataSourceId = await activeCollection.getPluginData(PLUGIN_COLLECTION_SYNC_REFERENCE_KEY)
+const syncSlugFieldId = await activeCollection.getPluginData(PLUGIN_COLLECTION_SYNC_SLUG_KEY)
+
+const syncDataSource = syncDataSourceId ? await getDataSources(syncDataSourceId) : null
+
+let savedFieldsConfig: FieldConfig[] | undefined
+
+if (syncDataSource) {
+    savedFieldsConfig = computeFieldConfigs(existingFields, syncDataSource)
+}
+
+if (framer.mode === "syncManagedCollection" && savedFieldsConfig && syncDataSource && syncSlugFieldId) {
+    try {
+        await syncCollection(
+            syncDataSource,
+            savedFieldsConfig.filter(field => field.field && !field.isNew),
+            syncSlugFieldId
+        )
+        await framer.closePlugin(`Synchronization successful`, {
+            variant: "success",
+        })
+    } catch (error) {
+        console.error(error)
+        framer.closePlugin(`Failed to sync collection`, {
+            variant: "error",
+        })
+    }
+}
```
Shouldn't all of these things only happen when the mode is `syncManagedCollection`? Otherwise its just going to delay the app from showing up, right?

jonastreub replied:
I think it would be much better to move all this into an async function

jonastreub replied:
And then the result of the function could be a string that says something like:

```ts
async function syncExistingCollection(): Promise<"success" | "needsConfiguration">
```

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,126 @@
+}
+
+export async function getDataSources(collection: string): Promise<DataSource> {
+    return await fetch(`/datasources/${collection}.json`).then(res => res.json())
```
Nit. Let's not use abbreviations. `res` → `result`

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,126 @@
+        if (existingField) {
+            newField = existingField
+        } else {
+            newField = {
+                id: fieldId,
+                name,
+                type: fieldType.type,
+                userEditable: false,
+            } as ManagedCollectionField
```
Do we need to cast this?

triozer replied:
Yes, because `type` could be `"enum" | "collectionReference" | "multiCollectionReference"` and then the field schema changes (either requiring `collectionId` or `cases`).

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+                    type="checkbox"
+                    disabled={isUnsupported}
+                    checked={!isDisabled}
+                    onChange={e => {
+                        e.stopPropagation()
+                    }}
```
Nit. Let's not use abbreviations: `e` → `event`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+                disabled={isDisabled}
+                placeholder={placeholder}
+                value={fieldName || ""}
+                onChange={e => {
```
`e` → `event`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+                value={fieldName || ""}
+                onChange={e => {
+                    if (!field) return
+                    if (!e.target.value.trim()) {
+                        onFieldNameChange(field.id, fieldConfig.name)
+                    } else {
+                        onFieldNameChange(field.id, e.target.value)
```
Should the field name always be trimmed? Also on line 66?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+    const isDisabled = useMemo(() => isUnsupported || isIgnored, [isUnsupported, isIgnored])
+
+    return (
+        <Fragment>
```
Are we using `<Fragment>` instead of `<>` on purpose?

triozer replied:
No, this is a remnant from when we split the component. We can update it to use the shorthand <> syntax

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+    const fieldName = hasFieldNameChanged ? field!.name : ""
+    const placeholder = isUnsupported ? "Unsupported Field" : fieldConfig.name
+
+    const isDisabled = useMemo(() => isUnsupported || isIgnored, [isUnsupported, isIgnored])
```
Why is this memoed?

triozer replied:
Also a remnant, before I was doing all the calculations in one memo but I forgot this one when moving over

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+        new Set(savedFieldsConfig?.filter(field => field.isNew).map(field => field.field!.id))
+    )
+
+    const possibleSlugFields = useMemo(
+        () =>
+            fields
+                .map(field => field.field)
+                .filter(isNotNull)
+                .filter(field => !disabledFieldIds.has(field.id) && field.type === "string"),
+        [fields, disabledFieldIds]
```
I think this would be easier to read if the inline function would use curly braces

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+    dataSource,
+    savedSlugFieldId,
+    onSubmit,
+}: {
+    savedFieldsConfig: FieldConfig[] | undefined
+    existingFields: ManagedCollectionField[]
+    dataSource: DataSource
+    savedSlugFieldId: string | null
+    onSubmit: (dataSource: DataSource, fields: FieldConfig[], slugFieldId: string) => Promise<void>
+}) {
```
I mostly prefer if component props are extracted into a standalone interface. Do you prefer inline like this?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+                        name="slugField"
+                        className="field-input"
+                        value={slugFieldId ?? ""}
+                        onChange={e => setSlugFieldId(e.target.value)}
```
`e` → `event`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+                    <hr className="sticky-top" />
+                    <button disabled={isSyncing}>
+                        {isSyncing ? (
+                            "Importing..."
```
`...` → `…`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,225 @@
+        )
+    }
+
+    const handleFieldToggle = (fieldId: string) => {
+        setDisabledFieldIds(current => {
+            const nextSet = new Set(current)
+            if (nextSet.has(fieldId)) {
+                nextSet.delete(fieldId)
+
+                // If we're re-enabling a string field and there's no valid slug field,
+                // set this field as the slug field
+                const field = fields.find(field => field.field?.id === fieldId)
+                if (field?.field?.type === "string") {
+                    const currentSlugField = fields.find(field => field.field?.id === slugFieldId)
+                    if (!currentSlugField || nextSet.has(slugFieldId ?? "")) {
+                        setSlugFieldId(fieldId)
+                    }
+                }
+            } else {
+                nextSet.add(fieldId)
+
+                // If the disabled field is the slug field, update it to the next
+                // possible slug field
+                if (fieldId === slugFieldId) {
+                    const nextSlugField = possibleSlugFields.find(field => field.id !== fieldId)
+                    if (nextSlugField) {
+                        setSlugFieldId(nextSlugField.id)
+                    }
+                }
+            }
```
I have a feeling that this can be simplified a lot. But my brain is too fried right now

triozer replied:
I did a pass on this!

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,166 @@
+    const fields = dataSource.fields
+
+    for (const [name, fieldType] of Object.entries(fields)) {
+        const fieldId = generateHash(name)
```
What's the reason for using a hash here?

triozer replied:
There is no reason to keep the hash anymore, this can be removed everywhere now, before we used to hash values because we were storing them.

jonastreub replied:
Great. Let's clean it up 🧹

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,166 @@
+            continue
+        }
+
+        const itemId = generateHash(slugValue)
```
Why use a hash and not just use the slug value?

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+const syncDataSourceId = await activeCollection.getPluginData(PLUGIN_KEYS.SYNC_REFERENCE)
+const syncSlugFieldId = await activeCollection.getPluginData(PLUGIN_KEYS.SYNC_SLUG)
+
+const result = await syncExistingCollection(activeCollection, syncDataSourceId, syncSlugFieldId)
```
Nice that this is now just a simple function call 👏 

I see that the result is used for many things that have nothing to do with syncing an existing collection. That makes it quite hard to understand what the result actually is

jonastreub replied:
I wonder if it would be simpler if `syncExistingCollection` would do nothing except for syncing the existing collection. Then that function is something you can forget about after reading it. You don't even need to know the implementation

triozer replied:
The only small thing doing so, is that in case the synchronization fails, we will re-fetch the data source when going opening the management view (we do open it automatically when sync fails, but we can also just close the plugin there).

I think it's the impact may be negligible enough to just allow it.

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+
+const activeCollection = await framer.getManagedCollection()
+
+const syncDataSourceId = await activeCollection.getPluginData(PLUGIN_KEYS.SYNC_REFERENCE)
+const syncSlugFieldId = await activeCollection.getPluginData(PLUGIN_KEYS.SYNC_SLUG)
```
What does the prefix _sync_ mean here?

triozer replied:
As in `configuredDataSourceId`, sync means that it has been imported.

jonastreub replied:
That's a much better name. The term _sync_ can mean too many things

jonastreub commented on examples/cms/src/main.tsx:
```diff
@@ -0,0 +1,14 @@
+import ReactDOM from "react-dom/client"
+import { App } from "./App.tsx"
+
+const root = document.getElementById("root")
+if (!root) throw new Error("Root element not found")
+
+ReactDOM.createRoot(root).render(
+    <React.StrictMode>
+        <App />
+    </React.StrictMode>
+)
```
I wonder if this would be a better place to add the logic for syncing an exists collection. Because then React is clearly completely skipped when nothing needs to be rendered. Wdyt?

triozer replied:
Nice one, will do!

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+                            value={selectedDataSourceId || ""}
+                        >
+                            <option value="" disabled>
+                                Choose...
```
Nit. In text it's better to use the `…` symbol instead of actual dots `...`

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+    const showFieldsMapping = async (event: React.FormEvent<HTMLFormElement>) => {
+        event.preventDefault()
+
+        if (!selectedDataSourceId) {
+            framer.notify("Please select a data source", { variant: "error" })
+            return
+        }
```
When can this happen?

triozer replied:
It can't really happen

jonastreub replied:
Then I would throw an error instead

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    if (selectedDataSource) {
+        return (
+            <FieldMapping
+                dataSource={selectedDataSource}
+                savedFieldsConfig={result.status === "needsConfiguration" ? result.savedFieldsConfig : null}
+                existingFields={existingFields}
+                savedSlugFieldId={syncSlugFieldId}
+                onSubmit={syncCollection}
+            />
+        )
```
This is nicely abstracted in a component, shall we do the same for data source selection? Then the parent component just decides one or the other. Would make it easier to see a clear overview of the different UI we render

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,134 @@
+                throw new Error("Failed to load data source")
+            }
+            setSelectedDataSource(dataSource)
+            setExistingFields(await activeCollection.getFields())
```
Can there be any fields yet?

triozer replied:
Nope!

jonastreub replied:
Let's then remove this line of code

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,166 @@
+    items: Record<string, unknown>[]
+}
+
+export interface FieldConfig {
+    name: string
+    isNew: boolean
+    field: ManagedCollectionField | null
```
A field already has a `name` attribute right? Why do we need a separate name in the config? Also, can't we compute `isNew` by comparing the fields on the node with the mutable fields in React state?

triozer replied:
This is the original name of the field, as set in the API

jonastreub replied:
I think it would be more clear if you have the current field and mutable fields as separate state. And then you can compute this instead of having multiple sources of truth

jonastreub commented on examples/cms/datasources/articles.json:
```diff
@@ -0,0 +1,49 @@
+{
+    "id": "articles",
+    "fields": {
+        "Title": {
+            "type": "string"
+        },
+        "Content": {
+            "type": "formattedText"
+        },
+        "Reading time": {
+            "type": "number"
+        },
+        "Featured": {
+            "type": "boolean"
+        }
```
Nit. Might be fine, but not very realistic that an API uses the same types as us. Maybe we should name the content type here `"html"` just to make that a little more clear

triozer replied:
We can! This was the first iteration, but it would add complexity mapping those values to ours.

I think @tom-james-watson @huntercaron may have stronger opinions

triozer replied:
For the sake of the example, I added it and a comment on "custom type"

huntercaron replied:
Interesting, this is not how I expected this to be customised 

For example, this is what i naturally started doing:

<img width="772" alt="Screenshot 2025-01-03 at 15 50 35" src="https://github.com/user-attachments/assets/fcea5159-a9b6-43f1-b406-6d001452f7d0" />

Then it stopped working pretty quick

triozer replied:
Yes because `computeFieldsFromDataSource` is missing a case for `link`.

jonastreub reviewed (commented):
Getting better. Left some comments. I can do another pass after you go over these

jonastreub commented:
Btw. The QA steps are looking strange. Github can't handle the lines of whitespace very well. I would just remove that whitespace

triozer commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    useLayoutEffect(() => {
+        framer.showUI({
+            width: UI_DEFAULTS.MAPPING_WIDTH,
+            height: UI_DEFAULTS.MAPPING_HEIGHT,
+            minWidth: UI_DEFAULTS.MAPPING_WIDTH,
+            minHeight: UI_DEFAULTS.MAPPING_HEIGHT,
+            resizable: true,
+        })
+    }, [])
+
```
To be removed

```suggestion
```

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,64 @@
+
+interface AppProps {
+    collection: ManagedCollection
+    dataSourceId: string | null
+    slugFieldId: string | null
```
Something that only becomes clear after reading the code is that `dataSourceId` and `slugFieldId` are the ones previously configured before the plugin was launched. I think we should make that clear in the naming of these props. That will also explain that these will never change at runtime

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,64 @@
+        collection.getFields().then(setExistingFields)
+    }, [collection])
+
+    useEffect(() => {
+        if (!dataSourceId) {
+            return
+        }
+
+        getDataSource(dataSourceId).then(setDataSource)
+    }, [dataSourceId])
+
+    if (!dataSource) {
+        return <SelectDataSource dataSources={getDataSourcesIds()} onSelectDataSource={setDataSource} />
```
When a `dataSourceId` is known, and that data source is loading, we currently render `SelectDataSource`. That seems wrong because you should never be able to change the data source after it has been selected, right?

jonastreub replied:
And what happens if the data source fails to load?

triozer replied:
> what happens if the data source fails to load?

Then you'll be shown the select datasource view

jonastreub replied:
But isn't that wrong? You can only select a data source once, right?

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,64 @@
+    const [dataSource, setDataSource] = useState<DataSource | null>(null)
+    const [existingFields, setExistingFields] = useState<ManagedCollectionField[]>([])
+
+    useLayoutEffect(() => {
+        if (!dataSource) {
+            framer.showUI({
+                width: UI_DEFAULTS.SETUP_WIDTH,
+                height: UI_DEFAULTS.SETUP_HEIGHT,
+                resizable: false,
+            })
+            return
+        }
+
+        framer.showUI({
+            width: UI_DEFAULTS.MAPPING_WIDTH,
+            height: UI_DEFAULTS.MAPPING_HEIGHT,
+            minWidth: UI_DEFAULTS.MAPPING_WIDTH,
+            minHeight: UI_DEFAULTS.MAPPING_HEIGHT,
+            resizable: true,
+        })
```
I think it would be better to make the size depends on whether a data source has been selected. Then it can render at the correct size during the initial render. So something like:

```ts
const hasDataSourceSelected = Boolean(dataSourceId || dataSource)
```

jonastreub replied:
And I would rename `dataSourceId` to something like `initialDataSourceId` or `previouslyConfiguredDataSourceId`

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,64 @@
+            <FieldMapping
+                collection={collection}
+                dataSource={dataSource}
+                existingFields={existingFields}
```
`existingFields` is only used within `FieldMapping`, shall we move the loading in there?

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+export function computeFieldsFromDataSource(dataSource: DataSource): ManagedCollectionField[] {
+    const fields: ManagedCollectionField[] = []
+
+    for (const [fieldId, field] of Object.entries(dataSource.fields)) {
+        let type = field.type
+
+        /**
+         * Here you can add support for other types, usually you would want to convert your APIs types to Framer types.
+         */
+
+        if (type === "html") {
+            type = "formattedText"
+        }
+
+        fields.push({
+            id: fieldId,
+            name: fieldId,
+            type,
+            userEditable: false,
+        } as ManagedCollectionField)
+    }
```
A cast like `as ManagedCollectionField` is bad practice. Because you lose some type safety. I think it's better to have more repetition than a cast. I think you can prevent doing the cast using a switch statement on the field type

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+
+export type ManagedCollectionFieldType = ManagedCollectionField["type"]
+
+type DataSourceField =
+    | {
+          type: ManagedCollectionFieldType
+      }
+    | {
+          type: "html"
+      }
```
With Vite, would it be possible to get these types from the JSON directly without having to manually specifying these?

triozer replied:
I don't think so as we will need to import (instead of fetching) the JSON here

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+    for (const item of dataSource.items) {
+        const slugValue = item[slugField.id]
+        if (typeof slugValue !== "string" || !slugValue) {
+            framer.notify(`Skipping item ${item.id} because it doesn't have a valid slug`, {
+                variant: "warning",
+            })
```
What is `item.id`?

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+        for (const [fieldName, value] of Object.entries(item)) {
+            const field = fields.find(field => field.id === fieldName)
+
+            // Field is in the data but should not be synced
+            if (!field) {
+                console.warn(`Skipping field ${fieldName} because it may have been ignored`)
+                continue
```
Nice. Let's also include the slug to make this more useful

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+                continue
+            }
+
+            // In a real-world scenario, we would need to convert the value to the correct type
+            fieldData[field.id] = value
```
What does this comment mean?

triozer replied:
It's indicating that the current implementation is simplified and doesn't perform type conversion of the values, but a real-world use case, the types may not directly match!

jonastreub replied:
Okay. I think this comment is adding more confusion then that its helpful. I think it would be useful to link to our documentation here, where we explain the expected values for each type. Because in most cases you probably don't need to do any conversion, also in real-world scenarios

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+    await collection.setPluginData(PLUGIN_KEYS.SLUG_FIELD_ID, slugFieldId)
+}
+
+type SyncResult = "success" | "needsSetup" | "needsConfiguration"
```
`needsSetup` and `needsConfiguration` sounds exactly the same. Only after reading the code I understand the difference. I would try to make these names more descriptive. Something like:
- needsDataSource
- needsFieldConfiguration

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,150 @@
+    try {
+        await syncCollection(collection, dataSource, existingFields, slugFieldId)
+        return "success"
+    } catch (error) {
+        console.error(error)
+        framer.notify(`Failed to sync collection`, {
+            variant: "error",
+        })
+        return "needsConfiguration"
```
Nice 👏

jonastreub commented on examples/cms/src/main.tsx:
```diff
@@ -0,0 +1,34 @@
+const configuredCollectionId = await activeCollection.getPluginData(PLUGIN_KEYS.COLLECTION_ID)
+const configuredSlugFieldId = await activeCollection.getPluginData(PLUGIN_KEYS.SLUG_FIELD_ID)
+
+const result = await syncExistingCollection(activeCollection, configuredCollectionId, configuredSlugFieldId)
```
Nit. I would name `result` → `syncResult`. Make the condition on line 17 easier. If we don't care about the other strings names of `syncExistingCollection` we could also make it return an object like so:

```ts
const { didSync } = await syncExistingCollection(activeCollection, configuredCollectionId, configuredSlugFieldId)

if (didSync) {
  await framer.closePlugin(`Synchronization successful`, { variant: "success" })
  return
}

// Render the UI
const root = document.getElementById("root")
…
```

jonastreub replied:
Why return an object and not just a boolean? Using the object the implementation of `syncExistingCollection` gets to decide how the boolean value is named. Otherwise it's easy to introduce a bug. For example when someone refactors the function to return `syncFailed` instead

jonastreub commented on examples/cms/src/components/Spinner.tsx:
```diff
@@ -0,0 +1,40 @@
+    inheritColor?: boolean
+}
+
+function styleForSize(size: SpinnerProps["size"]) {
+    switch (size) {
+        case "normal":
+            return styles.normalStyle
+        case "medium":
+            return styles.mediumStyle
+        case "large":
+            return styles.largeStyle
+    }
```
Nit. I would make the return type explicit here, that way you would get a type error if the size type is extended

triozer replied:
I'm not sure on how to do that as type of `styles.normalStyle` would be `string`

jonastreub replied:
```ts
function styleForSize(size: SpinnerProps["size"]): string {
```

Would this work?

triozer replied:
Oh, I thought we wanted something else than string

jonastreub replied:
Oh now, I just want to make sure you get a type error when adding a new size that isn't handled yet

jonastreub commented on examples/cms/src/constants.ts:
```diff
@@ -0,0 +1,11 @@
+    SLUG_FIELD_ID: "slugFieldId",
+} as const
+
+export const UI_DEFAULTS = {
+    SETUP_WIDTH: 320,
+    SETUP_HEIGHT: 305,
+    MAPPING_WIDTH: 360,
+    MAPPING_HEIGHT: 425,
```
Instead of SETUP and MAPPING I would use the same names as the components you use for rendering. Data source selection and field mapping. And now that all UI is controlled from the app component. Isn't it easier to just inline these values?

jonastreub commented on examples/cms/src/SelectDataSource.tsx:
```diff
@@ -0,0 +1,64 @@
+    return (
+        <main className="setup">
+            <div className="logo">
+                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 150">
+                    <path
+                        fill="#999"
+                        d="M75 33c18.778 0 34 7.611 34 17S93.778 67 75 67s-34-7.611-34-17 15.222-17 34-17Zm34 40.333C109 82.538 93.778 90 75 90s-34-7.462-34-16.667V60c0 9.389 15.222 17 34 17 18.776 0 33.997-7.61 34-16.997v13.33ZM109 84v.497c0-.166-.005-.332-.015-.497Zm0 13.333C109 106.538 93.778 114 75 114s-34-7.462-34-16.667V84h.015c-.01.166-.015.333-.015.5 0 9.113 15.222 16.5 34 16.5 18.776 0 3 ...
+                    />
+                </svg>
```
An inline SVG like this feel really heavy. And it's impossible to know what it represents by just reading this code. I would make it a nicely named component

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                    <span>Field</span>
+                    {fields.map((field, i) => (
+                        <FieldMappingRow
+                            key={field.id || i}
```
Is it possible for the field `id` to not exist? If not, I would remove the index fallback. If you do need the index fallback I would rename `i` → `index`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+}
+
+export function FieldMapping({ collection, existingFields, dataSource, slugFieldId, onImport }: FieldMappingProps) {
+    const originalFields = useMemo(() => computeFieldsFromDataSource(dataSource), [dataSource])
```
What does the term _original_ mean here? I guess a better name would be something like `possibleFieldsFromDataSource`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+    const originalFields = useMemo(() => computeFieldsFromDataSource(dataSource), [dataSource])
+    const [fields, setFields] = useState(mergeFieldsWithExistingFields(originalFields, existingFields))
+
+    const [disabledFieldIds, setDisabledFieldIds] = useState<Set<string>>(() => {
+        if (existingFields.length === 0) {
+            return new Set()
+        }
+
+        return new Set(
+            fields
+                .filter(field => !existingFields.find(existingField => existingField.id === field.id))
+                .map(field => field.id)
+        )
+    })
```
There is a lot going on here. I think this could be simplified by breaking it up a little bit.

```ts
const existingFieldIds = new Set(existingFields.map(getId))
const disabledIds = new Set<string>()
for (const field of fields) {
  const isExistingField = existingFieldIds.has(field.id)
  if (isExistingField) continue
  disabledIds.add(field.id)
}
return disabledIds
```

jonastreub replied:
Notice how this is much more readable. Mostly because its only doing a single thing per line

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+        )
+    })
+
+    const possibleSlugFields = useMemo(() => {
+        return fields.filter(field => {
+            const isStringType = field.type === "string"
+            const isEnabled = !disabledFieldIds.has(field.id)
+
+            return isStringType && isEnabled
+        })
+    }, [fields, disabledFieldIds])
```
Why does the possible slug field logic care if the field is disabled?

jonastreub replied:
I would expect you to be able to select any field of the valid type

triozer replied:
Previously (and I think in all our plugins too) we only allowed non-ignored fields to be used as slugs, but I agree we should allow all string fields to be slug candidates!

jonastreub replied:
Right, so instead just disable and check the checkbox if the field is the selected slug?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+        })
+    }, [fields, disabledFieldIds])
+
+    const [selectedSlugFieldId, setSelectedSlugFieldId] = useState<string | null>(
+        slugFieldId ?? possibleSlugFields[0]?.id ?? null
+    )
```
Renaming `slugFieldId` to something like `initialSlugFieldId` would make this much more clear

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+
+    const [isSyncing, setIsSyncing] = useState(false)
+
+    const handleFieldNameChange = (fieldId: string, name: string) => {
+        setFields(prev => prev.map(field => (field.id === fieldId ? { ...field, name } : field)))
+    }
```
Nit. `prev` → `previousFields`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                return updatedDisabledFieldIds
+            }
+
+            if (isEnabling && (!slugFieldId || prev.has(slugFieldId))) {
+                // When enabling a string field and there is no valid slug field, make it the slug
+                setSelectedSlugFieldId(fieldId)
+            } else if (!isEnabling && fieldId === slugFieldId) {
+                // When disabling the current slug field, find next available one
+                const nextSlugField = possibleSlugFields.find(f => f.id !== fieldId)
+                setSelectedSlugFieldId(nextSlugField?.id ?? null)
+            }
```
If we want to keep it working like this I think it would be much better to calculate the automatic slug field fallback int her render function instead of this callback. You can memoize that calculation if it's expensive

triozer replied:
I'd say it's fine to remove it now that we allow all source string fields to be slug candidates.

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,64 @@
+                dataSource={dataSource}
+                existingFields={existingFields}
+                slugFieldId={slugFieldId}
+                onImport={syncCollection}
```
Why is `onImport` a property, can't you call `syncCollection` from within the `FieldMapping` component?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+}
+
+function FieldMappingRow({ originalField, field, isIgnored, onFieldToggle, onFieldNameChange }: FieldMappingRowProps) {
+    const isUnsupported = !field
```
Is this possible?

triozer replied:
Nooooo

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+
+function FieldMappingRow({ originalField, field, isIgnored, onFieldToggle, onFieldNameChange }: FieldMappingRowProps) {
+    const isUnsupported = !field
+    const hasFieldNameChanged = field!.name !== originalField.name
+    const fieldName = hasFieldNameChanged ? field!.name : ""
```
The exclamation mark here makes it seems like field can not be undefined. Seems like unnecessary complexity

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                />
+                <span>{originalField.name}</span>
+            </div>
+            <svg xmlns="http://www.w3.org/2000/svg" width="8" height="16">
+                <path
+                    d="M 3 11 L 6 8 L 3 5"
+                    fill="transparent"
+                    strokeWidth="1.5"
+                    stroke="#999"
+                    strokeLinecap="round"
+                    strokeLinejoin="round"
+                ></path>
+            </svg>
```
What is this icon? Let's turn it into a component as well

triozer replied:
This is the chevron icon! I moved it into a component.

<img width="23" alt="image" src="https://github.com/user-attachments/assets/bf08d495-23f8-4652-ad22-7ce83d795c63" />

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                onChange={event => {
+                    if (!field) return
+
+                    const value = event.target.value
+                    if (!value.trim()) {
+                        onFieldNameChange(field.id, originalField.name)
+                    } else {
+                        onFieldNameChange(field.id, value.trimStart())
+                    }
```
Let's trim on line 64, so you don't need to do that on the other lines

triozer replied:
We need both `trim()` and `trimStart()` here:
  - trim() in the condition checks if the input is all whitespace
  - trimStart() in the value ensures no leading spaces

We can't just use `trim()` everywhere as it would prevent trailing spaces, which I think we want to allow (e.g., "Test A")

Alternatively, I can trim when calling `collection.setFields`?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                placeholder={placeholder}
+                value={fieldName || ""}
+                onChange={event => {
+                    if (!field) return
```
Not possible I think

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+
+function FieldMappingRow({ originalField, field, isIgnored, onFieldToggle, onFieldNameChange }: FieldMappingRowProps) {
+    const isUnsupported = !field
+    const hasFieldNameChanged = field!.name !== originalField.name
+    const fieldName = hasFieldNameChanged ? field!.name : ""
```
If the field name hasn't changed we render an empty string, doesn't that mean that the value can suddenly disappear when writing the same word as the field has?

Let's say the field is already named _Cat_, and you want to add a suffix, _Toy_, now when you start typing _Cat_ the input is suddenly cleared. Haven't tested this, just think so by reading the code…

triozer replied:
Yeah true! I updated the code, it got simplified a lot

jonastreub reviewed (commented):
Sorry if the amount of feedback seems overwhelming. Since this will be the starter kit for many customers I like it to become as simple and clear as we can possibly make it. And it's getting really good already 👏

triozer commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,244 @@
+                    type="checkbox"
+                    disabled={isUnsupported}
+                    checked={!isDisabled}
+                    onChange={event => {
+                        event.stopPropagation()
+                    }}
```
The stopPropagation() prevents the click event from reaching the parent div, which would otherwise trigger - once from the checkbox and once from the div's click handler.

jonastreub replied:
How is the checkbox change handled though? I don't see a change handler

jonastreub replied:
What I mean is that the checkbox input should be focusable by keyboard, but then using the keyboard to check or uncheck it would not do anything I guess, right? If so, that's bad accessibility

triozer replied:
Yeah, actually it works! I added a box shadow to outline the whole input when using the keyboard 🦾

jonastreub replied:
How does it work when you are not listening for events?

triozer replied:
When you interact with the checkbox (either by clicking or using keyboard Space/Enter):

- The browser fires a click event
- Our `onChange` handler only does stopPropagation() to prevent double-triggering
- The click event bubbles up to the parent div's handler which toggles the field (and updates isIgnored)
- The checkbox reflects this through `checked={!isIgnored}`

So while it might look like we're not handling the checkbox events, we are - just at the parent div level rather than directly on the checkbox. This pattern works well here since we want the entire row to be clickable anyway.

jonastreub replied:
TIL. I had no idea it worked like this. Very strange still

jonastreub replied:
Btw. I just checked in a code sandbox and I don't think the `stopPropagation` is needed. I tried both clicking the checkbox and outer div, and both resulted in a single click event.

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,86 @@
+            })
+    }, [previouslyConfiguredDataSourceId])
+
+    if (!isLoadingConfiguredDataSource && !dataSource) {
+        return <SelectDataSource dataSourceIds={getDataSourcesIds()} onSelectDataSource={setDataSource} />
+    }
+
+    if (isLoadingConfiguredDataSource) {
+        return <Spinner />
+    }
```
By moving the rendering of `Spinner`  up before `SelectDataSource`, you don't need to check for `!isLoadingConfiguredDataSource`

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,86 @@
+    }, [hasDataSourceSelected, dataSource])
+
+    useEffect(() => {
+        if (!previouslyConfiguredDataSourceId || hasStartedLoadingDataSource.current) {
+            return
+        }
```
Why do you need the logic for `hasStartedLoadingDataSource`? The value of `previouslyConfiguredDataSourceId` can't change at runtime, right?

triozer replied:
The ref is needed because in React's strict mode (development), useEffect runs twice by design. Without `hasStartedLoadingDataSource`, you'd fetch the data twice. While `previouslyConfiguredDataSourceId` won't change, the ref prevents this double-fetch.

jonastreub replied:
React is running the effect twice, to make sure you clean up your effects properly. They definitely don't want you to add workaround like these

jonastreub replied:
You can cancel the first fetch using an abort controller

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,86 @@
+    )
+    const hasDataSourceSelected = Boolean(isLoadingConfiguredDataSource || dataSource)
+
+    useLayoutEffect(() => {
+        framer.showUI({
+            width: hasDataSourceSelected ? 360 : 320,
+            height: hasDataSourceSelected ? 425 : 305,
+            minWidth: hasDataSourceSelected ? 360 : undefined,
+            minHeight: hasDataSourceSelected ? 425 : undefined,
+            resizable: dataSource !== null,
```
Any reason we use `hasDataSourceSelected` for all layout states except for `resizable`?

triozer replied:
Yes, plugins remember their previous window dimensions, so we need to be careful about when we allow resizing to prevent ending up with an oversized window showing just a small spinner!

`dataSource !== null` is specifically used for resizable because we only want resizing enabled when we actually have the data source loaded, not during the loading state

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,86 @@
+        return <Spinner />
+    }
+
+    if (dataSource) {
+        return (
+            <FieldMapping
+                collection={collection}
+                dataSource={dataSource}
+                initialSlugFieldId={previouslyConfiguredSlugFieldId}
+            />
+        )
+    }
+
+    assertNever(
+        `Invalid state: ${JSON.stringify({
+            previouslyConfiguredDataSourceId,
+            isLoadingConfiguredDataSource,
+            dataSource,
+        })}`
+    )
```
I think the `assertNever` looks way too scary for something that should never happen anyway. You already have an early return when there is no data source

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,86 @@
+    )
+}
+
+const assertNever = (message: string): never => {
+    throw new Error(message)
```
Let's remove this

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+    const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+        event.preventDefault()
+
+        if (!selectedSlugField) {
+            // This can't happen because the form will not submit if no slug field is selected
+            // but TypeScript can't infer that.
+            console.error("There is no slug field selected. Sync will not be performed")
```
Should we show a toast? I think user feedback would be useful

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+            <form onSubmit={handleSubmit}>
+                <label className="slug-field" htmlFor="slugField">
+                    Slug Field
+                    <select
+                        name="slugField"
+                        className="field-input"
+                        value={selectedSlugField ? selectedSlugField.id : ""}
+                        onChange={event =>
+                            setSelectedSlugField(
+                                possibleSlugFields.find(field => field.id === event.target.value) ?? null
+                            )
+                        }
+                        required
```
Nit. I would move attributes without a value like `required` all the way up, that way it's easier to read without them hanging in the void.

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+                        name="slugField"
+                        className="field-input"
+                        value={selectedSlugField ? selectedSlugField.id : ""}
+                        onChange={event =>
+                            setSelectedSlugField(
+                                possibleSlugFields.find(field => field.id === event.target.value) ?? null
+                            )
+                        }
```
This can be simplified by doing only one operation per line:

```ts
const selectedFieldId = event.target.value
const selectedField = possibleSlugFields.find(field => field.id === selectedFieldId)
if (!selectedField) return
setSelectedSlugField(selectedField)
```

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+            })
+    }, [collection, sourceFields, initialSlugFieldId])
+
+    const handleFieldNameChange = (fieldId: string, name: string) => {
+        setFields(previousFields => previousFields.map(field => (field.id === fieldId ? { ...field, name } : field)))
+    }
```
This works but it's hard to read. Let's do the same thing again, one operation per line.

```ts
const updatedFields = previousFields.map(field => {
  if (field.id !== fieldId) return field
  return {...field, name }
})
setFields(updatedFields)
```

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+    )
+})
+
+interface FieldMappingProps {
+    collection: ManagedCollection
+    dataSource: DataSource
+    initialSlugFieldId: string | null
+}
+
+type FieldMappingStatus = "mapping-fields" | "loading-fields" | "syncing-collection"
+
+export function FieldMapping({ collection, dataSource, initialSlugFieldId }: FieldMappingProps) {
```
Nit. Since the Props are so important to the component I would move that next to the component and have other types like `FieldMappingStatus` be further away. I think it would also be fine to have that type inline since it isn't used anywhere else

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+    const [fields, setFields] = useState<ManagedCollectionField[]>([])
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState<Set<string>>(new Set())
+
+    useEffect(() => {
+        collection
+            .getFields()
+            .then(fields => {
+                setFields(mergeFieldsWithExistingFields(sourceFields, fields))
+
+                if (Boolean(initialSlugFieldId) === false && fields.length === 0) {
+                    return
+                }
+
+                const ignoredFields = sourceFields.filter(
+                    field => !fields.some(existingField => existingField.id === field.id)
+                )
+
+                setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
+            })
+            .catch(error => {
+                console.error(error)
+                framer.notify("Failed to load existing fields.", { variant: "error" })
+            })
+            .finally(() => {
+                setStatus("mapping-fields")
+            })
+    }, [collection, sourceFields, initialSlugFieldId])
```
This effect can run multiple times and it is doing something async, so we need to be able to cancel it when the effect is cleaned up

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+        possibleSlugFields.find(field => field.id === initialSlugFieldId) ?? possibleSlugFields[0] ?? null
+    )
+
+    const [fields, setFields] = useState<ManagedCollectionField[]>([])
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState<Set<string>>(new Set())
```
I thin it would be good to mark these as readonly. And I wonder if the defaults should be lifted outside of rendering, so they don't have to be recreated on every render.

```ts
// Outside of rendering
type ManagedCollectionFields = readonly ManagedCollectionField[]
const initialManagedCollectionFields: ManagedCollectionFields = []

type FieldId = string
type FieldIds = ReadonlySet<FieldId>
const initialFieldIds: FieldIds = new Set()

// Within the component
const [fields, setFields] = useState<ManagedCollectionFields>(initialManagedCollectionFields)
const [ignoredFieldIds, setIgnoredFieldIds] = useState<FieldIds>(initialFieldIds)
```

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+    const [fields, setFields] = useState<ManagedCollectionField[]>([])
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState<Set<string>>(new Set())
+
+    useEffect(() => {
+        collection
+            .getFields()
+            .then(fields => {
+                setFields(mergeFieldsWithExistingFields(sourceFields, fields))
+
+                if (Boolean(initialSlugFieldId) === false && fields.length === 0) {
```
I would move the check if there is no `initialSlugFieldId` to the beginning of the effect, because that means we don't need to do any work

jonastreub replied:
And I don't think we need to check for a length of zero. Fine to set an empty array again. No reason to add extra conditionals for

jonastreub replied:
Oh wait. I missed that it's `&&`. So you can't do that

jonastreub replied:
I still don't understand this early return though

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+                    return
+                }
+
+                const ignoredFields = sourceFields.filter(
+                    field => !fields.some(existingField => existingField.id === field.id)
+                )
```
Let's name what's happening here by doing one operation per line with clearly named variables

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+                    field => !fields.some(existingField => existingField.id === field.id)
+                )
+
+                setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
```
If you create a utility named `getId` you can use it in a tone of places like this: `ignoredFields.map(getId)`

jonastreub replied:
We have a similar utility in Vekter

triozer replied:
Yeah, but we actually actually only done twice, not sure it is worth adding?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,233 @@
+            await syncCollection(
+                collection,
+                dataSource,
+                fields.filter(field => !ignoredFieldIds.has(field.id)),
```
Let's move this out into a variable, something like:

```ts
const fieldsToUpdate = fields.filter(field => !ignoredFieldIds.has(field.id))
```

jonastreub commented on examples/cms/src/SelectDataSource.tsx:
```diff
@@ -0,0 +1,73 @@
+import { Spinner } from "./components/Spinner"
+import { DataSource, getDataSource } from "./data"
+
+function Logo() {
+    return (
+        <div className="logo">
+            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 150 150">
+                <path
+                    fill="#999"
+                    d="M75 33c18.778 0 34 7.611 34 17S93.778 67 75 67s-34-7.611-34-17 15.222-17 34-17Zm34 40.333C109 82.538 93.778 90 75 90s-34-7.462-34-16.667V60c0 9.389 15.222 17 34 17 18.776 0 33.997-7.61 34-16.997v13.33Z"
+                />
+            </svg>
+        </div>
+    )
+}
```
I would make icon components just the SVG, no other HTML tags. Logo is a generic name, maybe the name of the component could be more descriptive so users will recognize it.

jonastreub commented on examples/cms/src/main.tsx:
```diff
@@ -0,0 +1,38 @@
+import { PLUGIN_KEYS } from "./constants"
+import { syncExistingCollection } from "./data"
+
+const activeCollection = await framer.getManagedCollection()
+
+const previouslyConfiguredDataSourceId = await activeCollection.getPluginData(PLUGIN_KEYS.DATA_SOURCE_ID)
+const previouslyConfiguredSlugFieldId = await activeCollection.getPluginData(PLUGIN_KEYS.SLUG_FIELD_ID)
+
+const { didSync } = await syncExistingCollection(
+    activeCollection,
+    previouslyConfiguredDataSourceId,
+    previouslyConfiguredSlugFieldId
+)
+
+if (didSync) {
+    await framer.closePlugin(`Synchronization successful`, {
+        variant: "success",
+    })
+} else {
+    const root = document.getElementById("root")
+    if (!root) throw new Error("Root element not found")
+
+    ReactDOM.createRoot(root).render(
+        <React.StrictMode>
+            <App
+                collection={activeCollection}
+                previouslyConfiguredDataSourceId={previouslyConfiguredDataSourceId}
+                previouslyConfiguredSlugFieldId={previouslyConfiguredSlugFieldId}
+            />
+        </React.StrictMode>
+    )
+}
```
This reads like a poem ❤️

jonastreub commented:
The icon doesn't seem centered

<img width="367" alt="image" src="https://github.com/user-attachments/assets/39151e89-e41d-4d07-8308-3233d629188e" />

jonastreub commented:
Since the left side acts as a button, I don't think the text should be selectable

<img width="247" alt="image" src="https://github.com/user-attachments/assets/5b0044fa-8657-45af-aeeb-e5dbf5dff61e" />

jonastreub commented:
When typing something in this field and pressing <kbd>Enter</kbd> the plugin syncs and automatically closes. That feels unexpected because I thought I was only updating the name.

<img width="441" alt="image" src="https://github.com/user-attachments/assets/0e454147-326a-4473-824d-e2272f09bc0b" />

jonastreub commented:
Would be nice if the reading time isn't the same for all items

<img width="220" alt="image" src="https://github.com/user-attachments/assets/e1833c7e-3001-490e-b4b7-64d0b0cb81a3" />

jonastreub commented:
Getting really good 💪

jonastreub commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,87 @@
+        return <Spinner />
+    }
+
+    if (!dataSource) {
+        return <SelectDataSource dataSourceIds={getDataSourcesIds()} onSelectDataSource={setDataSource} />
```
I would call `getDataSourcesIds` within the component `SelectDataSource`. No reason to make something static an API

jonastreub replied:
Als naming wise I would only make the last word plural. So:
`getDataSourcesIds` → `getDataSourceIds`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,223 @@
+
+interface FieldMappingRowProps {
+    field: ManagedCollectionField
+    isIgnored: boolean
```
Nit. When creating a variable in a `let` or `const` I like to prefix the variable by something like _is_ or _has_. But for property names I don't. That's because native HTML elements do that neither. It's `disabled` and `readonly`, and not `isDisabled` and `isReadOnly`

jonastreub replied:
Btw. Currently the Props are named `isIgnored` and `onFieldToggle`. Those two are related when you read the implementation but naming wise they seem unrelated

jonastreub replied:
I wonder if the term `enabled` or `disabled` is easier to understand btw

triozer replied:
Indeed, I'll rename to `disabled`. Keeping in mind that in future iteration we can have `unsupported` fields too.

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,223 @@
+    )
+})
+
+const managedCollection = await framer.getManagedCollection()
+const initialManagedCollectionFields: ManagedCollectionFields = await managedCollection.getFields()
+const initialFieldIds: FieldIds = new Set()
+
+interface FieldMappingProps {
+    collection: ManagedCollection
```
Why are you loading the `managedCollection` outside of rendering and then also passing it into the component as a property?

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,223 @@
+    const [fields, setFields] = useState(initialManagedCollectionFields)
+    const [ignoredFieldIds, setIgnoredFieldIds] = useState(initialFieldIds)
+
+    useEffect(() => {
+        setFields(mergeFieldsWithExistingFields(sourceFields, initialManagedCollectionFields))
+
+        const existingFieldIds = new Set(initialManagedCollectionFields.map(field => field.id))
+        const ignoredFields = sourceFields.filter(sourceField => !existingFieldIds.has(sourceField.id))
+
+        if (initialSlugFieldId) {
+            setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
+        }
+
+        setStatus("mapping-fields")
```
I think this used to be an effect because it depended on data that wasn't available yet. But currently this could be calculated outside of an effect I think. Probably because you are loading the entire collection and field outside the component?

jonastreub commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,180 @@
+                    id: fieldId,
+                    name: fieldId,
+                    type: "formattedText",
+                    userEditable: false,
+                })
+                break
+            case "string":
+            case "number":
+            case "boolean":
+            case "color":
+            case "formattedText":
+                fields.push({
+                    id: fieldId,
+                    name: fieldId,
+                    type: field.type,
+                    userEditable: false,
```
`userEditable` is optional and something we are probably going to deprecate, shall we just not include it in the example?

triozer replied:
It is not optional yet:
```ts
/**
 * Any kind of collection field definition that was created by a plugin and is
 * supported by the API.
 */
type ManagedCollectionField = SupportedCollectionField & {
    /** Is the user able to edit the field within the UI. */
    userEditable: boolean;
};
```

jonastreub replied:
Oh no. Let's make it optional ASAP

triozer replied:
Mhhh, actually, I think we should use the type

```ts
export type EditableManagedCollectionField = SupportedCollectionField &
    Partial<Pick<ManagedCollectionField, "userEditable">>
```

triozer replied:
But it is not exported by `framer-plugin`

jonastreub commented:
In Safari just clicking the field toggle button results in the text being selected, making it clear it isn't actually a button

<img width="137" alt="image" src="https://github.com/user-attachments/assets/a21dc9dd-caac-4204-a3bd-6bdf3ab09efe" />

jonastreub commented:
Because there is only a single plain text field its impossible to QA if the slug field selector works as expected when there are multiple options

jonastreub reviewed (commented):
Last few details. Getting really close now. Great work 👏

huntercaron commented on examples/cms/src/App.css:
```diff
@@ -0,0 +1,165 @@
+    grid-column: span 2 / span 2;
+}
+
+.setup,
+.mapping {
+    composes: no-scrollbar;
+}
```
lets just add the scrollbar class to the elements directly so its more obvious

huntercaron commented on examples/cms/src/App.css:
```diff
@@ -0,0 +1,165 @@
+}
+
+/* Chrome, Safari and Opera */
+.no-scrollbar::-webkit-scrollbar {
+    display: none;
+}
```
lets move this into framer.css as "framer-hide-scrollbar"

huntercaron commented on examples/cms/src/App.css:
```diff
@@ -0,0 +1,165 @@
+    scrollbar-width: none; /* Firefox */
+}
+
+[data-framer-theme="light"] .scroll-fade {
+    background: linear-gradient(to bottom, transparent, white);
+}
+
+[data-framer-theme="dark"] .scroll-fade {
+    background: linear-gradient(180deg, rgba(18, 18, 18, 0) 0%, rgb(17, 17, 17) 97.8%);
+}
```
into framer.css as "framer-scroll-fade-vertical"

huntercaron commented on examples/cms/src/App.css:
```diff
@@ -0,0 +1,165 @@
+    gap: 15px;
+}
+
+.mapping footer::before {
+    content: "";
+    position: absolute;
+    top: -45px;
+    left: 0;
+    width: 100%;
+    height: 45px;
+    background: linear-gradient(to bottom, transparent, var(--framer-color-bg));
+    pointer-events: none;
+}
```
not in use

triozer replied:
oh no, this is the one in use not the `.scroll-fade`

huntercaron commented on examples/cms/src/App.tsx:
```diff
@@ -0,0 +1,87 @@
+export function App({ collection, previouslyConfiguredDataSourceId, previouslyConfiguredSlugFieldId }: AppProps) {
+    const [dataSource, setDataSource] = useState<DataSource | null>(null)
+
+    const [isLoadingConfiguredDataSource, setIsLoadingConfiguredDataSource] = useState(
```
do we need to say "configured" each time, some of these lines become really long with these variable names and its harder to read

huntercaron commented on examples/cms/src/data.ts:
```diff
@@ -0,0 +1,180 @@
+    })
+}
+
+interface SyncResult {
+    didSync: boolean
+}
```
this is only used once, lets just put it inline?

huntercaron commented:
<img width="375" alt="Screenshot 2024-12-31 at 11 01 18" src="https://github.com/user-attachments/assets/f7c747f0-723b-40e6-bd63-746d8b42d048" />

inbetween the two states there is this inbetween size that makes it feel jumpy

triozer commented:
> In Safari just clicking the field toggle button results in the text being selected, making it clear it isn't actually a button

@jonastreub fixed this by making it a real button (will benefit of accessibility too)

> Because there is only a single plain text field its impossible to QA if the slug field selector works as expected when there are multiple options

I added a new plain text field to categories!

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,235 @@
+interface FieldMappingRowProps {
+    field: ManagedCollectionField
+    disabled: boolean
+    onDisable: (fieldId: string) => void
```
`onDisable`, sounds like it will only be used to disable the field. I would name it `onToggleDisabled`

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,235 @@
+
+                setFields(mergeFieldsWithExistingFields(sourceFields, collectionFields))
+
+                const existingFieldIds = new Set(collectionFields.map(field => field.id))
+                const ignoredFields = sourceFields.filter(sourceField => !existingFieldIds.has(sourceField.id))
+
+                if (initialSlugFieldId) {
+                    setIgnoredFieldIds(new Set(ignoredFields.map(field => field.id)))
+                }
```
I guess this if condition might be hard to understand. Can we just check if the existing field ids has a size of zero? Would that be easier to understand? Either way we should probably add a comment

triozer replied:
We can't, because when syncing for the first time `existingFields` would be empty.

Also this iteration allows to sync even if there is no field selected, if there is a valid slug selected then syncing will work!

jonastreub commented on examples/cms/src/FieldMapping.tsx:
```diff
@@ -0,0 +1,235 @@
+        })
+    }
+
+    const handleFieldDisable = (fieldId: string) => {
```
`handleFieldDisable` → `toggleFieldDisabledState`

jonastreub reviewed (approved):
Those were the last few nits code wise. Stellar work 👏

jonastreub commented:
Just QA'd and it works flawlessly now ❤️

triozer commented:
> inbetween the two states there is this inbetween size that makes it feel jumpy

cc @huntercaron using `useEffect` instead of `useLayoutEffect` seems to fix this!

huntercaron commented on examples/cms/src/SelectDataSource.tsx:
```diff
@@ -0,0 +1,76 @@
+
+            <form onSubmit={handleSubmit}>
+                <label htmlFor="collection">
+                    Collection
```
this text should be text secondary color not text tertiary

huntercaron commented:
> > inbetween the two states there is this inbetween size that makes it feel jumpy
> 
> cc @huntercaron using `useEffect` instead of `useLayoutEffect` seems to fix this!

Its closer! 

There is still one frame of weird layout. Perhaps some overflow

<img width="426" alt="Screenshot 2025-01-02 at 12 46 26" src="https://github.com/user-attachments/assets/ba11d8c9-83b8-4cba-80a0-005074de2f54" />

Sometimes it shows a spinner in the middle as well? This should feel more solid. 


https://github.com/user-attachments/assets/50d3a90d-cc5d-42c6-b6ac-3e6bc4168da8

huntercaron commented on examples/cms/package.json:
```diff
@@ -0,0 +1,35 @@
+    "preview": "vite preview",
+    "pack": "npx framer-plugin-tools@latest pack"
+  },
+  "dependencies": {
+    "framer-plugin": "^2",
+    "motion": "^11.13.4",
+    "react": "^18",
+    "react-dom": "^18",
+    "vite-plugin-mkcert": "^1"
+  },
+  "devDependencies": {
+    "@eslint/js": "^9",
+    "@types/react": "^18",
+    "@types/react-dom": "^18",
+    "@vitejs/plugin-react": "^4.3.1",
+    "@vitejs/plugin-react-swc": "^3",
+    "eslint": "^9.9.0",
+    "eslint-plugin-react-hooks": "^5.1.0-rc.0",
+    "eslint-plugin-react-refresh": "^0.4.9",
+    "globals": "^15.9.0",
+    "typescript": "^5.3",
+    "typescript-eslint": "^8.0.1",
+    "vite": "^5",
+    "vite-plugin-framer": "^1"
```
Are we using all of these? is that rc version something we want to be using?

triozer replied:
I think we can remove globals and update `eslint-plugin-react-hooks`

huntercaron commented on package-lock.json:
Do we need to commit the lockfile? I think no as we want people to be able to use the package manager of their choice

triozer replied:
This is the workspace lockfile, so people will still be able to use another package manager when cloning.

huntercaron replied:
Ah sorry! of course yes

huntercaron commented:
Could you update to vite 6? we use that in the default template now

huntercaron commented:
Lets make sure to use nice quotes here https://smartquotesforsmartpeople.com/

<img width="770" alt="Screenshot 2025-01-03 at 14 26 00" src="https://github.com/user-attachments/assets/a705c0ba-2d90-4597-9418-604885001ffc" />


Also should we allow people to unselect _everything_? it won't work anyways so I would imagine we'd want to disable the button here when no valid slug fields are selected or when you deselect the active slug? I remember we discussed some similar stuff but I cannot recall at the moment

<img width="492" alt="Screenshot 2025-01-03 at 14 27 25" src="https://github.com/user-attachments/assets/4580b94b-79ac-44b3-958c-a64e8a37b9c3" />

triozer commented:
> Also should we allow people to unselect everything?

I think yes!

> it won't work anyways 

This is a bug. I just fixed

> so I would imagine we'd want to disable the button here when no valid slug fields are selected or when you deselect the active slug? I remember we discussed some similar stuff but I cannot recall at the moment

So we removed this logic, now the slug field is unrelated to the fields you are mapping. So the possible slugs will always be the list of of all plain-text fields.

cc @huntercaron

huntercaron commented:
> This is a bug. I just fixed
> 
> >

👌 ok that works too

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,174 @@
+
+type DataSourceField =
+    | {
+          type: ManagedCollectionFieldType
```
What is the HMTL option here for? hard to follow by just reading it

triozer replied:
For more context; see https://github.com/framer/plugins/pull/134#discussion_r1896644257.

Should I add more comments around the types?

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,174 @@
+import { framer } from "framer-plugin"
+import { PLUGIN_KEYS } from "./constants"
+
+export type ManagedCollectionFieldType = ManagedCollectionField["type"]
+export type ManagedCollectionFields = readonly ManagedCollectionField[]
+export type FieldIds = ReadonlySet<ManagedCollectionField["id"]>
```
This type is only used once, it can be inlined

huntercaron replied:
I am not a big fan of creating these types. It adds another layer of abstraction away from the core APIs and should be up to the author to decide if they want to do this.

We should only do this if its saving a lot of lines of code

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,174 @@
+}
+
+export async function getDataSource(dataSourceId: string, abortSignal?: AbortSignal): Promise<DataSource> {
+    return await fetch(`/datasources/${dataSourceId}.json`, { signal: abortSignal }).then(response => response.json())
```
I'm assuming this function is where I should add most of my logic? maybe worth a little comment on what that looks like?

huntercaron commented on starters/cms/datasources/categories.json:
```diff
@@ -0,0 +1,36 @@
+{
+    "id": "categories",
+    "fields": {
+        "Title": {
+            "type": "string"
+        },
+        "Description": {
+            "type": "string"
+        },
+        "Color": {
+            "type": "color"
+        }
+    },
```
where does the structure for this come from? its just something we made up for this starter?

triozer replied:
Yes, it's made up.

huntercaron replied:
ok

huntercaron commented:
Why am I getting this?

<img width="543" alt="Screenshot 2025-01-03 at 15 34 52" src="https://github.com/user-attachments/assets/cb16cab3-7bb6-423a-81c6-a882313ba096" />

triozer commented on starters/cms/src/SelectDataSource.tsx:
```diff
@@ -0,0 +1,75 @@
+}
+
+export function SelectDataSource({ onSelectDataSource }: SelectDataSourceProps) {
+    const [dataSourceIds] = useState(getDataSourceIds())
```
> is the idea it will eventually be dynamic? seems like a bit much.

yes, this will ultimately be dynamic!

> I am also prettyyyy sure invoking the function like that without a closure will create a new value every render? but a full time engie would know better?

oooops, you are right it should be `useState(() => getDataSourceIds())`

huntercaron commented:
Trying to actually swap in my own cms source (bluesky) its not the most clear process still.

In theory it should be pretty simple, to get the data this is enough, but then I run into complexity of not sure how to set things up
```tsx
import { AtpAgent } from "@atproto/api"

  const agent = new AtpAgent({
    service: "https://public.api.bsky.app",
  })
  const profile = await agent.getProfile({ actor: "framer.com" })

  const { data } = await agent.getAuthorFeed({
    actor: profile.data.did,
    filter: "posts_and_author_threads",
    limit: 30,
  })
  
 ```

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,182 @@
+}
+
+export async function getDataSource(dataSourceId: string, abortSignal?: AbortSignal): Promise<DataSource> {
+    const dataSource: {
+        id: string
+        fields: {
+            name: string
+            type: string
+        }[]
+        items: Record<string, unknown>[]
+    } = await fetch(`/datasources/${dataSourceId}.json`, { signal: abortSignal }).then(response => response.json())
```
nit: lets split this on a few lines, its a bit weird to deconstruct directly from the fetch like that on the same line I think

huntercaron replied:
I also think keeping the DataSource type separate is easier to read rather than inline like this

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,182 @@
+    const fields: ManagedCollectionField[] = []
+    for (const field of dataSource.fields) {
+        switch (field.type) {
+            case "html":
+                fields.push({
+                    id: field.name,
+                    name: field.name,
+                    type: "formattedText",
+                    userEditable: false,
+                })
+                break
```
I think we can just remove this if its not being used. this is more of an example for the docs/readme than something that should be included in every file I think

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,182 @@
+            case "collectionReference":
+            case "multiCollectionReference":
+                console.warn(
+                    `Support for field type "${field.type}" is not implemented in this plugin. The field will be ignored.`
```
Support for “${field.type}” is not implemented in this Plugin

huntercaron commented on starters/cms/src/data.ts:
```diff
@@ -0,0 +1,182 @@
+                )
+                break
+            default: {
+                console.warn(`Unknown field type "${field.type}". The field will be ignored.`)
```
Unknown field type “${field.type}”

is enough I think

huntercaron commented:
ohhh I see now its a bit different as DataSource means it has the ID? but I want an inline type here without the ID?
<img width="495" alt="Screenshot 2025-01-03 at 17 51 16" src="https://github.com/user-attachments/assets/f5de03df-9142-4cfc-9aa3-807cb52e9dcf" />

triozer commented:
> ohhh I see now its a bit different as DataSource means it has the ID? but I want an inline type here without the ID?

Yes, we can do that but I think it's better if we re-use the types we already defined right?

huntercaron commented on starters/cms/src/constants.ts:
we don't need another file for this, lets move into data.ts

huntercaron commented on starters/cms/src/constants.ts:
we don't need another file for this, lets move into data.ts

triozer commented on starters/cms/vite.config.ts:
```diff
@@ -0,0 +1,15 @@
+    build: {
+        target: "ES2022",
+    },
+    server: {
+        cors: true,
+    },
```
Must be removed by bumping `vite-plugin-framer` version
